Experiment_Launcher_code/ExperimentManager.py

```diff
@@ -69,6 +71,22 @@ def __init__(self, comport,mouse_id,opponenttype):
         self.time_to_return_to_center = 0
 
         self.timestamps={}##for the video writer
+        self.event_lock = Lock()
+        self.trialevents=0
+
+    def on_press(self, key):
+            with self.event_lock:
+                if key == Key.esc:
+                    print("Experiment stopped by user.")
+                    self.trialevents = Events.ExperimentStopped.value
```

Comment:
This will not work. You are zeroing the trialevents variable at the beginning of the experiment loop. You will loose the key press. Create a different self.userStop bool  flag. set it to False at the beginning of the experiment and set it to True here. Check it in the experiment loop. If the flag is set, set the event.

Reply:
i tried it out and it seemed to be working .are you saying i should use a self.userStop bool flag as an alternative to the event_lock?

Reply:
So, ....
I missed a line in your code, so I did not see the other half of the Lock() usage. My bed.
Still, I don't like the usage of Lock() here, its not clean, its not needed, and it only seems to be working :-). I mean it will work most of the times ...
It's complicated to explain in writing.  Let's talk about it tomorrow.